test/controllers/users_controller_test.rb

```diff
@@ -0,0 +1,64 @@
+require "test_helper"
+
+describe UsersController do
+  let(:user) { users(:hana) }
+
+  describe "login_form" do 
+    it "can get the login page" do 
+      get login_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "login" do 
+    it "will login a returning user" do 
+      user_data = {
+        user: {
+          username: user.username,
+        },
+      }
+
+      expect {
+        post login_path, params: user_data
+      }.wont_change "User.count"
+
+      expect(session[:user_id]).must_equal user.id
+      expect(flash[:success]).must_equal "Successfully logged in as an existing user #{user.username}"
+    end
+
+    it "will create new user" do 
+      user_data = {
+        user: {
+          username: "new user",
```

Comment:
Good work getting both of the success cases here. There is also a failure case you're missing: what if the user can't be created? For example, if the username is "".